artifact/sha/sha.cpp

```diff
@@ -0,0 +1,126 @@
+#include <artifact/sha/sha.hpp>
+
+#include <common/common.hpp>
+
+#include <cstdint>
+#include <string>
+#include <ctime>
+#include <iomanip>
+#include <vector>
+#include <sstream>
+#include <algorithm>
+
+namespace mender {
+namespace sha {
+
+namespace log = mender::common::log;
+
+
+const ErrorCategoryClass ErrorCategory = ErrorCategoryClass();
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "ShaSumErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	InitializationError:
+		return "Initialization error";
+	ShasumCreationError:
+		return "Shasum creation error";
+	ShasumMismatchError:
+		return "Shasum mismatch error";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code) {
+	return error::Error(error_condition(code, ErrorCategory), "");
+}
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+Reader::Reader(io::Reader &reader) :
+	Reader::Reader {reader, ""} {
+}
+
+Reader::Reader(io::Reader &reader, std::string expected_sha = "") :
+	sha_handle_(EVP_MD_CTX_new(), [](EVP_MD_CTX *ctx) { EVP_MD_CTX_free(ctx); }),
+	wrapped_reader_ {reader},
+	buffer_(1024),
+	expected_sha_ {expected_sha} {
+	if (EVP_DigestInit_ex(sha_handle_.get(), EVP_sha256(), nullptr) != 1) {
+		log::Error("Failed to initialize the shasummer");
+		initialized = false;
+		return;
+	}
+	initialized = true;
+}
+
+expected::ExpectedSize Reader::Read(vector<uint8_t> &dst) {
+	if (!initialized) {
+		return MakeError(InitializationError);
+	}
+
+	auto bytes_read = wrapped_reader_.Read(buffer_);
+	if (!bytes_read) {
+		return bytes_read;
+	}
+
+	// bytes_read == 0 == EOF marker in our Reader/Writer interface implementation
+	if (bytes_read.value() == 0) {
+		auto expected_sha = ShaSum();
+		if (!expected_sha) {
+			return expected_sha.error();
+		}
+		auto actual_sha = expected_sha.value();
```

Comment:
`expected_sha` is a really confusing name here.

Reply:
Lol, yes. You are only using `actual_sha` twice, one of which is an error path, so maybe just rename `expected_sha` to `actual_sha`, and use it directly with the `value()` function.

Reply:
Yes, horrible naming. What about something like `stream_sha`?
* `calculated_sha`
* `real_sha`